Doc/spectramerger (#265)
OpenMS/pyopenms-docs: docs/source/algorithms.rst

    gf.filterExperiment(exp)
    # MzMLFile().store("test.filtered.mzML", exp)

Spectra Merge Algorithm
*************************

Spectra merge algorithms merge several spectra to e.g., improve spectrum quality by increasing S/N ratio.
Merging is usually done block wise (for MS1 and above) or (for MS2 and above) based on similar precursors.

> In any case, the number of scans will be reduced. Some methods allow to smooth or average spectra. 

Reviewer: is this true? you show examples below were it doesn't get reduced and the averaging also doesn't seem to reduce the number of spectra

.. code-block:: python

    # retrieve data 
    gh = "https://raw.githubusercontent.com/OpenMS/pyopenms-extra/master"
    urlretrieve(gh + "/src/data/small.mzML", "test.mzML")

    # load MS data and store as MSExperiment object
    exp = MSExperiment()
    MzMLFile().load('test.mzML', exp)

    spectra = exp.getSpectra()
    spectra_ms1 = [s for s in spectra if s.getMSLevel() == 1]
    print(f'Number of MS1 spectra before merge are {len(spectra_ms1)}')

    # merges blocks of MS1
    merger = SpectraMerger()

    merger.mergeSpectraBlockWise(exp)

    spectraMerged = exp.getSpectra()
    spectra_ms1 = [s for s in spectra if s.getMSLevel() == 1]
    print(f'Number of MS1 spectra after merge are {len(spectra_ms1)}')

.. code-block:: output 
  
    Number of MS1 spectra before merge are 183

    Number of MS1 spectra after merge are 37

Per default, the method ``mergeSpectraBlockWise`` of SpectraMerger merges 5 consectutive MS1 spectra into a block. Before the merge, we had 183 MS1 spectra. Now, we have 37 MS1 spectra left. 

The modified data structure can be stored on disk:

.. code-block:: python

    MzMLFile().store("mergedBlockWise.mzML", exp)

> SpectraMerger includes the method ``mergeSpectraPrecursors`` which allows the merging of spectra with similar precursors. The data must have MS2 spectra for the precursor information. 

Reviewer: ```suggestion
SpectraMerger includes the method ``mergeSpectraPrecursors`` which allows the merging of spectra with similar precursors. Note: the data needs to contain MS2 spectra with annotated precursor information.
```

.. code-block:: python 

    # load MS data and store as MSExperiment object
    exp = MSExperiment()
    MzMLFile().load('test.mzML', exp)

    spectra = exp.getSpectra()

    # spectra with ms_level = 2
    spectra_ms2 = [s for s in spectra if s.getMSLevel() == 2]
    print(f'Number of MS2 spectra before merge are {len(spectra_ms2)}')

    # merge spectra with similar precursors 
    merger = SpectraMerger()
    merger.mergeSpectraPrecursors(exp)

    spectraMerged = exp.getSpectra()
    spectra_ms2 = [s for s in spectra if s.getMSLevel() == 2]
    print(f'Number of MS2 spectra after merge are {len(spectra_ms2)}')

    # store modified data 
    # MzMLFile().store("mergedSimiPrecursors.mzML", exp)

.. code-block:: output

    Number of MS2 spectra before merge are 53

    Number of MS2 spectra after merge are 53

> We see that the number of MS2 spectra before and after the merge do not change. This means that the hierarchical clustering with single linkage of the basic LC-MS feature (here only RT and M/Z of the precursors) did not produce any clusters (blocks to merge). 

Reviewer: hmm note a great example if nothing gets merged. But let's keep it for now.

> We see that the number of MS2 spectra before and after the merge do not change. This means that the hierarchical clustering with single linkage of the basic LC-MS feature (here only RT and M/Z of the precursors) did not produce any clusters (blocks to merge). 

Reviewer: ```suggestion
In this example the number of MS2 spectra before and after the merge do not change. This means that precursors were to dissimilar in RT and m/z. 
```

> SpectraMerger presents a method ``average`` to average experimental data over neighbouring spectra. The block of neighbouring spectra depends on the averaging type: ``gaussian`` or ``tophat``. The gaussian type checks for a weight < cutoff value, whereas tophat averages over a range (by default 5 steps left and right from each scan). Per default SpectraMerger averages MS1 spectra. 

Reviewer: what are the steps here?

Author: One step is just one scan/ spectrum. So it would be like a window of 5 scans left and 5 scans right for each scan with averaging type tophat.

Reviewer: ah, thanks. I think then we should just talk about ± 5 neighboring spectra. The user might get confused if she reads steps.

    gf.filterExperiment(exp)
    # MzMLFile().store("test.filtered.mzML", exp)

Spectra Merge Algorithm
*************************

> The pyOpenMS spectra merge algorithms allows the merge of several spectra by either increasing S/N ratio (for MS1 and above) or merging scans which stem from similar precursors (for MS2 and above). In any case, the number of scans will be reduced. In addition, pyOpenMS provides the method to average spectra. 

Reviewer: Maybe "In addition, pyOpenMS provides the method to average spectra" is a bit unclear. I don't know the details here.

> The pyOpenMS spectra merge algorithms allows the merge of several spectra by either increasing S/N ratio (for MS1 and above) or merging scans which stem from similar precursors (for MS2 and above). In any case, the number of scans will be reduced. In addition, pyOpenMS provides the method to average spectra. 

Reviewer: ```suggestion
The pyOpenMS spectra merge algorithms allows the merge of several spectra by either increasing S/N ratio (for MS1 and above) or merging scans which stem from similar precursors (for MS2 and above). In any case, the number of scans will be reduced. Some of the methods available for merging average spectra. 
```

Reviewer 2: merging average spectra?

Different spectra merge algorithms are available in pyOpenMS:

- mergeSpectraBlockWise
- mergeSpectraPrecursors
- average (over neighbouring spectra for MS1 or above)

In order to have a better picture of the algorithms, each one will be introduced by an example below. We will use another example file compared to above and start with an example of merging spectra blockwise:

.. code-block:: python

    # retrieve data 
    gh = "https://raw.githubusercontent.com/OpenMS/pyopenms-extra/master"
    urlretrieve(gh + "/src/data/small.mzML", "test.mzML")

    # load MS data and store as MSExperiment object
    exp = MSExperiment()
    MzMLFile().load('test.mzML', exp)

    spectra = exp.getSpectra()
    print(f"There are {len(spectra)} spectra before merging.\n")

    # merges blocks of MS or MS2 spectra
    sm = SpectraMerger()
    sm.mergeSpectraBlockWise(exp)

    spectraMB = exp.getSpectra()
    print(f"There are {len(spectraMB)} spectra after merging")

.. code-block:: output 
  
    There are 236 spectra before merging.

>     There are 90 spectra after merging. 

Reviewer 2: Why 90? What defines the block size?

Author: > Thanks! But I can already see people asking the following questions. I think this needs a bit more detail.

Thanks for the reply! I wasn't sure how much I should go into detail. But okay, I will add more information.

As predicated, the number of scans decreased after merging MS spectra. The modified data structure can be stored on disk:

.. code-block:: python

    MzMLFile().store("mergedBlockWise.mzML", exp)

The merging of spectra with similar precursors progresses likewise:

.. code-block:: python 

    # load MS data and store as MSExperiment object
    exp = MSExperiment()
    MzMLFile().load('test.mzML', exp)

>     # merge spectra with similar precursors 

Reviewer 2: How is similar defined?

Author: Similar to the processing steps, I think that I should add a more detailed information block in each example for better understanding.

Reviewer: I think he means when are pcs considered similar (m/z ? Charge?)

    sm = SpectraMerger()
    sm.mergeSpectraPrecursors(exp)

    # store modified data 
    # MzMLFile().store("mergedBlockWise.mzML", exp)

``SpectraMerger`` presents a method to average experimental data over neighbouring spectra. The averaging types to be used by the method are either ``gaussian`` or ``tophat``:

.. code-block:: python 

    # load MS data and store as MSExperiment object
    exp = MSExperiment()
    MzMLFile().load('test.mzML', exp)

    # average spectra with gaussian
    sm = SpectraMerger()

>     sm.average(exp, "gaussian")  

Reviewer 2: What is the default number of spectra being merged?
Why no before/after comparison here?

Author: There were no changes in the spectra numbers with the example file.

Reviewer 2: haha ok. Doesn't that sound like something is wrong? Please check if this does what it is supposed to do.

Author: Yes, as @timosachsenberg suggested, the outcome of the average depends on the MS level. There is no change in the spectra number at MS2 level, whereas the total number of spectra changes to 183 for the MS1 level.

Reviewer 2: I am just looking for simple explanations: e.g.
Per default, SpectraMerger merges MS1. Before we had 6 MS1 spectra. Now we have 2 MS1 spectra, because per default SpectraMerger always merges 3 consecutive spectra.
Or:
The number of resulting spectra cannot be predicted since the algorithm combines spectra based on property XYZ.

Author: Thank you for the feedback! Then I will take this into account.

Reviewer: I did not check how complex the code is but if feasible you can take a look and find out what the differences are between the different modes and extend the documentation in the header file.

Author: The code is not very complex, but it takes a bit of time due to the different functions, but it's still manageable. I am also currently reading the code to understand the differences between each method better. Yes, I can do that.
